x-pack/plugins/security_solution/scripts/endpoint/common/agent_downloads_service.ts

```diff
@@ -79,59 +84,67 @@ class AgentDownloadStorage extends SettingsStorage<AgentDownloadStorageSettings>
     };
   }
 
+  /**
+   * Downloads the agent and stores it locally. Reuses existing downloads if available.
+   */
   public async downloadAndStore(
     agentDownloadUrl: string,
     agentFileName?: string
   ): Promise<DownloadedAgentInfo> {
-    this.log.debug(`Downloading and storing: ${agentDownloadUrl}`);
-
-    // TODO: should we add "retry" attempts to file downloads?
+    this.log.debug(`Starting download: ${agentDownloadUrl}`);
 
     await this.ensureExists();
-
     const newDownloadInfo = this.getPathsForUrl(agentDownloadUrl, agentFileName);
 
-    // If download is already present on disk, then just return that info. No need to re-download it
+    // Return cached version if the file already exists
     if (fs.existsSync(newDownloadInfo.fullFilePath)) {
       this.log.debug(`Download already cached at [${newDownloadInfo.fullFilePath}]`);
       return newDownloadInfo;
     }
 
-    try {
-      const outputStream = fs.createWriteStream(newDownloadInfo.fullFilePath);
-
-      await handleProcessInterruptions(
-        async () => {
-          const { body } = await nodeFetch(agentDownloadUrl);
-          await finished(body.pipe(outputStream));
-        },
-        () => {
-          fs.unlinkSync(newDownloadInfo.fullFilePath);
+    const maxAttempts = 2;
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      try {
+        this.log.info(
+          `Downloading agent from [${agentDownloadUrl}] to [${newDownloadInfo.fullFilePath}]`
+        );
+        const outputStream = fs.createWriteStream(newDownloadInfo.fullFilePath);
+
+        await handleProcessInterruptions(
+          async () => {
+            const { body } = await nodeFetch(agentDownloadUrl);
+            await finished(body.pipe(outputStream));
+          },
+          () => fs.promises.unlink(newDownloadInfo.fullFilePath) // Clean up on interruption
```

Comment:
is this `fs.promises.` valid? I can't find it in the docs for nodeJS. My understanding was that if you wanted to work with promise versions of methods you had to explicitly import them.

just use `unlink` since that was imported from the FS promises package

Reply:
reverted 👍